### PR TITLE
Replace matmul with einsum

### PR DESCRIPTION
**Pull request by jonb377**
Replace matmul with einsum in llama2 modeling code

### CODE REVIEW

File: src/transformers/models/llama/modeling_llama.py [hunk 1]
```diff
@@ -439,7 +444,12 @@ def forward(
 
         # upcast attention to fp32
         attn_weights = nn.functional.softmax(attn_weights, dim=-1, dtype=torch.float32).to(query_states.dtype)
-        attn_output = torch.matmul(attn_weights, value_states)
+        # attn_weights Batch Num_head Seq Kv_seq
+        # value_states Batch Num_head Seq Head_dim
+        # attn_output = torch.matmul(attn_weights, value_states)
+        assert attn_weights.shape == torch.Size((bsz, self.num_heads, q_len, kv_seq_len)), f'incorrect atten_weight shape: {attn_weights.shape}'
+        assert value_states.shape == torch.Size((bsz, self.num_heads, q_len, self.head_dim)), f'incorrect value_states shape: {value_states.shape}'
+        attn_output = torch.einsum('bnsk,bnsh->bnsh', attn_weights, value_states)
```
Comment (alanwaketan) [inline src/transformers/models/llama/modeling_llama.py]: Why there is a transpose here as well since the original code doesn't?

Comment (jonb377) [reply src/transformers/models/llama/modeling_llama.py]: This should be `bnsk,bnkh->bnsh`

**Comment by alanwaketan** [comment]
This change seems helping SPMD so I'm going to approve it.

**Review by alanwaketan — APPROVED** [review]
LGTM.

**Review by lsy323 — APPROVED** [review]
LGTM

**Pull request merged**